File: backend/render_thumbnails.py
#!/usr/bin/env python3
"""
Generate thumbnail images from GLB models using trimesh and pyrender
"""

import os
import sys
import json
import boto3
from botocore.client import Config

# Load environment variables
from dotenv import load_dotenv
load_dotenv()

AWS_ACCESS_KEY = os.getenv('AWS_ACCESS_KEY_ID')
AWS_SECRET_KEY = os.getenv('AWS_SECRET_ACCESS_KEY')
AWS_ENDPOINT = os.getenv('AWS_ENDPOINT', 's3.eu-west-1.wasabisys.com')
AWS_REGION = os.getenv('AWS_REGION', 'eu-west-1')
S3_BUCKET = os.getenv('S3_BUCKET', 'jigu1')

TEMP_DIR = 'temp_models'
THUMB_DIR = 'reference_images'

# Create directories
os.makedirs(TEMP_DIR, exist_ok=True)
os.makedirs(THUMB_DIR, exist_ok=True)

def get_s3_client():
    """Create S3 client for Wasabi"""
    return boto3.client(
        's3',
        endpoint_url=f'https://{AWS_ENDPOINT}',
        aws_access_key_id=AWS_ACCESS_KEY,
        aws_secret_access_key=AWS_SECRET_KEY,
        region_name=AWS_REGION,
        config=Config(signature_version='s3v4')
    )

def list_glb_models(s3):
    """List all GLB models in the bucket"""
    print("Fetching GLB models from S3...")
    response = s3.list_objects_v2(Bucket=S3_BUCKET, MaxKeys=500)
    
    models = []
    for obj in response.get('Contents', []):
        key = obj['Key']
        if key.lower().endswith('.glb'):
            # Skip person/human models
            if 'man' not in key.lower() and 'woman' not in key.lower() and 'person' not in key.lower():
                models.append(key)
    
    print(f"Found {len(models)} GLB models")
    return models

def list_existing_thumbnails(s3):
    """List existing thumbnails in S3"""
    print("Checking existing thumbnails...")
    try:
        response = s3.list_objects_v2(Bucket=S3_BUCKET, Prefix='reference_images/', MaxKeys=500)
        thumbs = []
        for obj in response.get('Contents', []):
            key = obj['Key']
            if key.lower().endswith(('.jpg', '.jpeg', '.png')):
                base = os.path.splitext(os.path.basename(key))[0]
                thumbs.append(base)
        print(f"Found {len(thumbs)} existing thumbnails")
        return thumbs
    except:
        return []

def download_model(s3, key):
    """Download a GLB model from S3"""
    local_path = os.path.join(TEMP_DIR, os.path.basename(key))
    print(f"  Downloading {key}...")
    s3.download_file(S3_BUCKET, key, local_path)
    return local_path

def upload_thumbnail(s3, local_path, s3_key):
    """Upload thumbnail to S3"""
    print(f"  Uploading to {s3_key}...")
    s3.upload_file(local_path, S3_BUCKET, s3_key, ExtraArgs={'ContentType': 'image/png'})

def render_thumbnail_trimesh(glb_path, output_path):
    """Render thumbnail using trimesh"""
    try:
        import trimesh
        import numpy as np
        from PIL import Image
        
        # Load the GLB file
        scene = trimesh.load(glb_path)
        
        # Get the scene or mesh
        if isinstance(scene, trimesh.Scene):
            # Combine all meshes
            meshes = []
            for name, geom in scene.geometry.items():
                if isinstance(geom, trimesh.Trimesh):
                    meshes.append(geom)
            if meshes:
                mesh = trimesh.util.concatenate(meshes)
            else:
                raise ValueError("No meshes found in scene")
        else:
            mesh = scene
        
        # Try to render using pyrender if available
        try:
            import pyrender
            
            # Create scene
            pr_scene = pyrender.Scene(bg_color=[240, 240, 240, 255])
            
            # Add mesh
            pr_mesh = pyrender.Mesh.from_trimesh(mesh)
            pr_scene.add(pr_mesh)
            
            # Add camera
            camera = pyrender.PerspectiveCamera(yfov=np.pi / 3.0)
            
            # Position camera to see the whole model
            bounds = mesh.bounds
            center = mesh.centroid
            size = np.max(bounds[1] - bounds[0])
            camera_distance = size * 2
            
            camera_pose = np.eye(4)
            camera_pose[:3, 3] = center + [0, 0, camera_distance]
            pr_scene.add(camera, pose=camera_pose)
            
            # Add light
            light = pyrender.DirectionalLight(color=[1.0, 1.0, 1.0], intensity=3.0)
            pr_scene.add(light, pose=camera_pose)
            
            # Render
            renderer = pyrender.OffscreenRenderer(256, 256)
            color, _ = renderer.render(pr_scene)
            renderer.delete()
            
            # Save image
            img = Image.fromarray(color)
            img.save(output_path)
            return True
            
        except ImportError:
            # Fallback: create a simple visualization
            print("  pyrender not available, using simple render...")
            
            # Create a simple top-down view
            bounds = mesh.bounds
            center = mesh.centroid
            
            # Project vertices to 2D
            vertices = mesh.vertices - center
            
            # Simple orthographic projection (front view)
            x = vertices[:, 0]
            y = vertices[:, 1]
            
            # Normalize to image coordinates
            x_min, x_max = x.min(), x.max()
            y_min, y_max = y.min(), y.max()
            
            scale = min(200 / (x_max - x_min + 0.001), 200 / (y_max - y_min + 0.001))
            
            x_img = ((x - x_min) * scale + 28).astype(int)
            y_img = (256 - ((y - y_min) * scale + 28)).astype(int)
            
            # Create image
            img = Image.new('RGB', (256, 256), (240, 240, 240))
            pixels = img.load()
            
            # Draw points
            for xi, yi in zip(x_img, y_img):
                if 0 <= xi < 256 and 0 <= yi < 256:
                    pixels[xi, yi] = (50, 50, 50)
            
            img.save(output_path)
            return True
            
    except Exception as e:
        print(f"  Render error: {e}")
        return False

def create_placeholder_thumbnail(glb_path, output_path):
    """Create a simple placeholder thumbnail"""
    try:
        from PIL import Image, ImageDraw, ImageFont
        
        base_name = os.path.splitext(os.path.basename(glb_path))[0]
        
        # Determine color based on filename
        color = (128, 128, 128)  # default gray
        name_lower = base_name.lower()
        
        if 'black' in name_lower: color = (26, 26, 26)
        elif 'gold' in name_lower or 'yellow' in name_lower: color = (255, 215, 0)
        elif 'silver' in name_lower or 'metal' in name_lower: color = (192, 192, 192)
        elif 'red' in name_lower or 'bloody' in name_lower: color = (220, 20, 60)
        elif 'blue' in name_lower: color = (65, 105, 225)
        elif 'green' in name_lower: color = (34, 139, 34)
        elif 'pink' in name_lower or 'rose' in name_lower: color = (255, 105, 180)
        elif 'white' in name_lower: color = (245, 245, 245)
        elif 'brown' in name_lower: color = (139, 69, 19)
        elif 'orange' in name_lower: color = (255, 140, 0)
        elif 'purple' in name_lower: color = (147, 112, 219)
        elif 'round' in name_lower: color = (100, 100, 100)
        elif 'cat' in name_lower: color = (80, 80, 80)
        elif 'aviator' in name_lower: color = (60, 60, 60)
        
        # Create image
        img = Image.new('RGB', (256, 256), (240, 240, 240))
        draw = ImageDraw.Draw(img)
        
        # Draw glasses shape
        # Left lens
        draw.ellipse([40, 90, 120, 160], outline=color, width=4)
        # Right lens
        draw.ellipse([136, 90, 216, 160], outline=color, width=4)
        # Bridge
        draw.line([120, 125, 136, 125], fill=color, width=3)
        # Left temple
        draw.line([40, 125, 10, 100], fill=color, width=2)
        # Right temple
        draw.line([216, 125, 246, 100], fill=color, width=2)
        
        # Add name
        try:
            font = ImageFont.truetype("arial.ttf", 10)
        except:
            font = ImageFont.load_default()
        
        short_name = base_name[:25] + '...' if len(base_name) > 25 else base_name
        draw.text((128, 200), short_name, fill=(100, 100, 100), anchor="mm", font=font)
        
        img.save(output_path)
        return True
        
    except Exception as e:
        print(f"  Placeholder error: {e}")
        return False

def main():
    print("=== GLB Thumbnail Generator ===\n")
    
    s3 = get_s3_client()
    
    models = list_glb_models(s3)
    existing = list_existing_thumbnails(s3)
    
    # Find models without thumbnails
    to_process = [m for m in models if os.path.splitext(os.path.basename(m))[0] not in existing]
    
    print(f"\nModels needing thumbnails: {len(to_process)}")
    
    if not to_process:
        print("All models have thumbnails!")
        return
    
    # Check if trimesh is available
    try:
        import trimesh
        use_trimesh = True
        print("Using trimesh for rendering")
    except ImportError:
        use_trimesh = False
        print("trimesh not available, using placeholder images")
        print("Install with: pip install trimesh pillow")
    
    processed = 0
    failed = 0
    
    for i, model_key in enumerate(to_process):
        base_name = os.path.splitext(os.path.basename(model_key))[0]
        print(f"\n[{i+1}/{len(to_process)}] {base_name}")
        
        try:
            # Download model
            local_glb = download_model(s3, model_key)
            thumb_path = os.path.join(THUMB_DIR, f"{base_name}.png")
            
            # Render thumbnail
            success = False
            if use_trimesh:
                success = render_thumbnail_trimesh(local_glb, thumb_path)
            
            if not success:
                success = create_placeholder_thumbnail(local_glb, thumb_path)
            
            if success and os.path.exists(thumb_path):
                # Upload to S3
                upload_thumbnail(s3, thumb_path, f"reference_images/{base_name}.png")
                print("  ✅ Success")
                processed += 1
            else:
                print("  ❌ Failed to create thumbnail")
                failed += 1
            
            # Cleanup
            if os.path.exists(local_glb):
                os.remove(local_glb)
                
        except Exception as e:
            print(f"  ❌ Error: {e}")
            failed += 1
    
    print(f"\n=== Summary ===")
    print(f"Processed: {processed + failed}")
    print(f"Success: {processed}")
    print(f"Failed: {failed}")
    
    if processed > 0:
        print("\nBuilding AI embeddings...")
        os.system("python match.py --build")

if __name__ == "__main__":
    main()

File: backend/render_thumbnails.mjs
import 'dotenv/config';
import AWS from 'aws-sdk';
import fs from 'fs';
import path from 'path';
import { execSync } from 'child_process';

const s3Endpoint = process.env.AWS_ENDPOINT || 's3.eu-west-1.wasabisys.com';
const s3Region = process.env.AWS_REGION || 'eu-west-1';
const BUCKET = process.env.S3_BUCKET || 'jigu1';

const s3 = new AWS.S3({
    endpoint: `https://${s3Endpoint}`,
    region: s3Region,
    accessKeyId: process.env.AWS_ACCESS_KEY_ID,
    secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY,
    s3ForcePathStyle: true,
    signatureVersion: 'v4'
});

// Create directories
const TEMP_DIR = 'temp_models';
const THUMB_DIR = 'reference_images';

if (!fs.existsSync(TEMP_DIR)) fs.mkdirSync(TEMP_DIR);
if (!fs.existsSync(THUMB_DIR)) fs.mkdirSync(THUMB_DIR);

async function listGLBModels() {
    console.log('Fetching GLB models from S3...');
    const data = await s3.listObjectsV2({ Bucket: BUCKET, MaxKeys: 500 }).promise();
    const glbFiles = (data.Contents || [])
        .filter(f => f.Key.toLowerCase().endsWith('.glb'))
        .filter(f => !f.Key.toLowerCase().includes('man') && !f.Key.toLowerCase().includes('woman') && !f.Key.toLowerCase().includes('person'))
        .map(f => ({ key: f.Key, size: f.Size }));
    console.log(`Found ${glbFiles.length} GLB models`);
    return glbFiles;
}

async function checkExistingThumbnails() {
    console.log('Checking existing thumbnails in S3...');
    try {
        const data = await s3.listObjectsV2({ 
            Bucket: BUCKET, 
            Prefix: 'reference_images/',
            MaxKeys: 500 
        }).promise();
        const thumbs = (data.Contents || [])
            .filter(f => f.Key.match(/\.(jpg|jpeg|png)$/i))
            .map(f => path.basename(f.Key, path.extname(f.Key)));
        console.log(`Found ${thumbs.length} existing thumbnails`);
        return thumbs;
    } catch (e) {
        return [];
    }
}

async function downloadModel(key) {
    const localPath = path.join(TEMP_DIR, path.basename(key));
    console.log(`  Downloading ${key}...`);
    const data = await s3.getObject({ Bucket: BUCKET, Key: key }).promise();
    fs.writeFileSync(localPath, data.Body);
    return localPath;
}

async function uploadThumbnail(localPath, s3Key) {
    console.log(`  Uploading thumbnail to ${s3Key}...`);
    const body = fs.readFileSync(localPath);
    await s3.upload({
        Bucket: BUCKET,
        Key: s3Key,
        Body: body,
        ContentType: 'image/png'
    }).promise();
}

async function main() {
    console.log('=== GLB Thumbnail Generator ===\n');
    
    const models = await listGLBModels();
    const existingThumbs = await checkExistingThumbnails();
    
    // Find models without thumbnails
    const modelsToProcess = models.filter(m => {
        const baseName = path.basename(m.key, '.glb');
        return !existingThumbs.includes(baseName);
    });
    
    console.log(`\nModels needing thumbnails: ${modelsToProcess.length}`);
    
    if (modelsToProcess.length === 0) {
        console.log('All models have thumbnails!');
        return;
    }
    
    // Check if we have the renderer
    const rendererPath = path.join(process.cwd(), 'thumbnail_renderer.mjs');
    if (!fs.existsSync(rendererPath)) {
        console.log('\nCreating thumbnail renderer...');
        createRenderer();
    }
    
    console.log('\n--- Processing Models ---\n');
    
    let processed = 0;
    let failed = 0;
    
    for (const model of modelsToProcess) {
        const baseName = path.basename(model.key, '.glb');
        console.log(`[${processed + failed + 1}/${modelsToProcess.length}] ${baseName}`);
        
        try {
            // Download model
            const localGlb = await downloadModel(model.key);
            
            // Render thumbnail using the renderer script
            const thumbPath = path.join(THUMB_DIR, `${baseName}.png`);
            
            try {
                execSync(`node thumbnail_renderer.mjs "${localGlb}" "${thumbPath}"`, {
                    timeout: 60000,
                    stdio: 'pipe'
                });
                
                if (fs.existsSync(thumbPath)) {
                    // Upload to S3
                    await uploadThumbnail(thumbPath, `reference_images/${baseName}.png`);
                    console.log(`  ✅ Success`);
                    processed++;
                } else {
                    console.log(`  ❌ Thumbnail not created`);
                    failed++;
                }
            } catch (renderErr) {
                console.log(`  ❌ Render failed: ${renderErr.message}`);
                failed++;
            }
            
            // Cleanup
            if (fs.existsSync(localGlb)) fs.unlinkSync(localGlb);
            
        } catch (err) {
            console.log(`  ❌ Error: ${err.message}`);
            failed++;
        }
        
        // Progress update every 10 models
        if ((processed + failed) % 10 === 0) {
            console.log(`\n--- Progress: ${processed} success, ${failed} failed ---\n`);
        }
    }
    
    console.log('\n=== Summary ===');
    console.log(`Total processed: ${processed + failed}`);
    console.log(`Successful: ${processed}`);
    console.log(`Failed: ${failed}`);
    
    // Build embeddings if we have new thumbnails
    if (processed > 0) {
        console.log('\nBuilding AI embeddings...');
        try {
            execSync('python match.py --build', { stdio: 'inherit' });
            console.log('✅ Embeddings built successfully!');
        } catch (e) {
            console.log('⚠️ Could not build embeddings automatically. Run: python match.py --build');
        }
    }
}

function createRenderer() {
    const rendererCode = `
import * as THREE from 'three';
import { GLTFLoader } from 'three/examples/jsm/loaders/GLTFLoader.js';
import fs from 'fs';
import path from 'path';

// Headless rendering requires node-canvas or similar
// For now, we'll create a simple placeholder

const args = process.argv.slice(2);
if (args.length < 2) {
    console.error('Usage: node thumbnail_renderer.mjs <input.glb> <output.png>');
    process.exit(1);
}

const inputPath = args[0];
const outputPath = args[1];

// Since we can't do headless WebGL easily in Node.js,
// we'll create a colored placeholder based on the filename
const baseName = path.basename(inputPath, '.glb').toLowerCase();

// Determine color based on filename
let color = '#808080'; // default gray
if (baseName.includes('black')) color = '#1a1a1a';
else if (baseName.includes('gold') || baseName.includes('yellow')) color = '#ffd700';
else if (baseName.includes('silver') || baseName.includes('metal')) color = '#c0c0c0';
else if (baseName.includes('red') || baseName.includes('bloody')) color = '#dc143c';
else if (baseName.includes('blue')) color = '#4169e1';
else if (baseName.includes('green')) color = '#228b22';
else if (baseName.includes('pink') || baseName.includes('rose')) color = '#ff69b4';
else if (baseName.includes('white')) color = '#f5f5f5';
else if (baseName.includes('brown')) color = '#8b4513';
else if (baseName.includes('orange')) color = '#ff8c00';
else if (baseName.includes('purple')) color = '#9370db';

// Create a simple SVG placeholder
const svg = \`<?xml version="1.0" encoding="UTF-8"?>
<svg width="256" height="256" xmlns="http://www.w3.org/2000/svg">
  <rect width="256" height="256" fill="#f0f0f0"/>
  <text x="128" y="128" text-anchor="middle" font-family="Arial" font-size="12" fill="\${color}">
    \${path.basename(inputPath, '.glb').substring(0, 20)}
  </text>
  <ellipse cx="80" cy="120" rx="40" ry="30" fill="none" stroke="\${color}" stroke-width="3"/>
  <ellipse cx="176" cy="120" rx="40" ry="30" fill="none" stroke="\${color}" stroke-width="3"/>
  <line x1="120" y1="120" x2="136" y2="120" stroke="\${color}" stroke-width="3"/>
  <line x1="40" y1="120" x2="20" y2="100" stroke="\${color}" stroke-width="2"/>
  <line x1="216" y1="120" x2="236" y2="100" stroke="\${color}" stroke-width="2"/>
</svg>\`;

// For now, just write a marker file - we need a proper renderer
fs.writeFileSync(outputPath.replace('.png', '.svg'), svg);
console.log('Created SVG placeholder (PNG rendering requires additional setup)');
process.exit(0);
`;
    
    fs.writeFileSync('thumbnail_renderer.mjs', rendererCode);
}

main().catch(console.error);
